fendermint/actors/gas_market/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright 2021-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use fil_actors_runtime::actor_error;
+use fil_actors_runtime::runtime::{ActorCode, Runtime};
+use fil_actors_runtime::SYSTEM_ACTOR_ADDR;
+use fil_actors_runtime::{actor_dispatch, ActorError};
+use fvm_ipld_encoding::tuple::*;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::METHOD_CONSTRUCTOR;
+use lazy_static::lazy_static;
+use num_derive::FromPrimitive;
+use std::cmp::Ordering;
+
+#[cfg(feature = "fil-actor")]
+fil_actors_runtime::wasm_trampoline!(EIP1559GasMarketActor);
+
+/// Base fee max change denominator as defined in [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559)
+const BASE_FEE_MAX_CHANGE_DENOMINATOR: u64 = 8;
+/// Elasticity multiplier as defined in [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559)
+const ELASTICITY_MULTIPLIER: u64 = 2;
+lazy_static! {
+    /// Initial base fee as defined in [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559)
+    static ref INITIAL_BASE_FEE: TokenAmount = TokenAmount::from_atto(1_000_000_000);
+    static ref MINIMAL_BASE_FEE: TokenAmount = TokenAmount::from_atto(100);
+}
+pub const IPC_GAS_MARKET_ACTOR_NAME: &str = "gas_market";
+pub type Gas = u64;
+pub type GasMarketReading = EIP1559GasState;
+
+#[derive(Serialize_tuple, Deserialize_tuple, Debug, Clone)]
+pub struct EIP1559GasState {
+    pub block_gas_limit: Gas,
+    pub base_fee: TokenAmount,
+}
+
+#[derive(Serialize_tuple, Deserialize_tuple, Debug, Clone)]
+pub struct BlockGasUtilization {
+    pub block_gas_used: Gas,
+}
+
+#[derive(Serialize_tuple, Deserialize_tuple, Debug, Clone)]
+pub struct SetConstants {
+    pub block_gas_limit: Gas,
+}
+
+pub struct EIP1559GasMarketActor {}
+
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum Method {
+    Constructor = METHOD_CONSTRUCTOR,
+    CurrentGasReading = frc42_dispatch::method_hash!("CurrentGasReading"),
+    SetConstants = frc42_dispatch::method_hash!("SetConstants"),
+    UpdateUtilization = frc42_dispatch::method_hash!("UpdateUtilization"),
+}
+
+impl EIP1559GasMarketActor {
+    /// Creates the actor
+    pub fn constructor(rt: &impl Runtime, st: EIP1559GasState) -> Result<(), ActorError> {
```

Comment:
This should take a ConstructorParams object, otherwise you're leaking the state object externally, which is highly undesirable as it breaks abstraction.

```rust
struct ConstructorParams {
    constants: Constants,
    initial_base_fee: TokenAmunt,
}